solidity/ecdsa/test/WalletRegistry.Upgrade.test.ts

```diff
@@ -0,0 +1,336 @@
+import { deployments, ethers, upgrades, helpers } from "hardhat"
+import chai, { expect } from "chai"
+import chaiAsPromised from "chai-as-promised"
+import { keccak256 } from "ethers/lib/utils"
+
+import { params, walletRegistryFixture } from "./fixtures"
+import { fakeRandomBeacon } from "./utils/randomBeacon"
+import { noMisbehaved, signAndSubmitCorrectDkgResult } from "./utils/dkg"
+import ecdsaData from "./data/ecdsa"
+import { createNewWallet } from "./utils/wallets"
+
+import type { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers"
+import type { WalletRegistry, WalletRegistryV2 } from "../typechain"
+import type { FactoryOptions } from "hardhat/types"
+import type { Contract } from "ethers"
+import type { UpgradeProxyOptions } from "@openzeppelin/hardhat-upgrades/src/utils/options"
+
+const { mineBlocksTo } = helpers.time
+const { AddressZero } = ethers.constants
+
+chai.use(chaiAsPromised)
+
+describe("WalletRegistry - Upgrade", async () => {
+  let proxyAdminOwner: SignerWithAddress
+  let EcdsaInactivity: Contract
+
+  before(async () => {
+    proxyAdminOwner = await ethers.getNamedSigner("esdm")
+    EcdsaInactivity = await ethers.getContract("EcdsaInactivity")
+  })
+
+  describe("upgradeProxy", () => {
+    describe("when new contract fails upgradeability validation", () => {
+      describe("when a variable was added before old variables", () => {
+        it("should throw an error", async () => {
+          await deployments.fixture()
+
+          await expect(
+            upgradeProxy("WalletRegistry", "WalletRegistryV2Invalid", {
+              factoryOpts: {
+                libraries: { EcdsaInactivity: EcdsaInactivity.address },
+                signer: proxyAdminOwner,
+              },
+              proxyOpts: {
+                constructorArgs: [AddressZero, AddressZero],
+                unsafeAllowLinkedLibraries: true,
+              },
+            })
+          ).to.be.rejectedWith(Error, "New storage layout is incompatible")
+        })
+      })
+
+      describe("when a variable was removed", () => {
+        it("should throw an error", async () => {
+          await deployments.fixture()
+
+          await expect(
+            upgradeProxy("WalletRegistry", "WalletRegistryV2MissingSlot", {
+              factoryOpts: {
+                libraries: { EcdsaInactivity: EcdsaInactivity.address },
+                signer: proxyAdminOwner,
+              },
+              proxyOpts: {
+                constructorArgs: [AddressZero, AddressZero],
+                unsafeAllowLinkedLibraries: true,
+              },
+            })
+          ).to.be.rejectedWith(
+            Error,
+            "Deleted `_maliciousDkgResultNotificationRewardMultiplier`"
+          )
+        })
+      })
+    })
```

Comment:
Non-blocking this PR, but I would love to see similar tests with a library.

I know, it should work the same. But just in case.

Reply:
Oh yes, I now see a TODO at line 287! ❤️